File: c/p25.c
/* p25.c - Project Euler problem 25
 * Blake Mitchell, 2012
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "eutil.h"

#define DIGCOUNT 1000

main()
{
        char *a, *b, *tmp;
        int term;
        int max = DIGCOUNT+1;

        a = (char*)malloc(max);
        b = (char*)malloc(max);
        tmp = (char*)malloc(max);
        
        inttostr(0, a, max);
        inttostr(1, b, max);

        for (term = 0; a[DIGCOUNT-1] == '0' || a[DIGCOUNT-1] == '\0'; term++) {
                memcpy(tmp, b, max);
                addstr_inplace(a, b, max);
                memcpy(a, tmp, max);
        }
        printf("The first Fibonacci number with %d digits is:\n\n", DIGCOUNT);
        strnum_reverse(a, max);
        printf("%s\n", a);
        printf("\nThis is term %d in the sequence\n", term);
        
        free(a);
        free(b);
        free(tmp);
}

File: go/p25.go
// p25.go - Project Euler problem 25
// Blake Mitchell, 2012

package main


import (
        "fmt"
        "math/big"
)


func main() {
        bfib := BigFibonacci()
        term, n := bfib()
        for ; len(n.String()) < 1000; term, n = bfib() { }
        fmt.Println("The first Fibonacci number with 1000 digits is:\n")
        fmt.Println(n)
        fmt.Printf("\nThis is term %d in the sequence.\n", term)
}


func BigFibonacci() func() (int, *big.Int) {
        a := big.NewInt(int64(0))
        b := big.NewInt(int64(1))
        tmp := big.NewInt(0)
        term := 0
        return func() (int, *big.Int) {
                tmp.Add(b, big.NewInt(int64(0)))
                b.Add(a, b)
                a.Add(tmp, big.NewInt(int64(0)))
                term++
                return term, a
        }
}


